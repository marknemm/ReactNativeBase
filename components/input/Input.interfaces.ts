import type { FormFieldProps, ValidateFn, ValidationRules } from '@interfaces/form';
import type { InputProps as RneInputProps } from '@rneui/themed';
import type { PropsWithChildren, ReactNode, Ref } from 'react';
import type { FieldPath, FieldValues, Message, Path, ValidationRule } from 'react-hook-form';
import type { TextInput } from 'react-native';
import type { Mask } from 'react-native-mask-input';
import type Input from './Input';

/**
 * The {@link Input} component properties.
 *
 * @param TFieldValues The type of the form data.
 * @param TContext The type of the form context.
 * @param TFieldName The form field name.
 * @extends InputProps The {@link InputProps} from the `@rneui/themed` package.
 * @extends FormFieldProps The {@link FormFieldProps} from the `@interfaces/form` package.
 * @extends ValidationRules The {@link ValidationRules} from the `@interfaces/form` package.
 */
export interface InputProps<
  TFieldValues extends FieldValues = any,
  TContext = any,
  TFieldName extends FieldPath<TFieldValues> = Path<TFieldValues>
> extends Omit<RneInputProps, 'maxLength'>,
  FormFieldProps<TFieldValues, TContext, TFieldName>,
  ValidationRules<TFieldValues, TFieldName>
{

  /**
   * The mask to apply to the {@link Input}.
   */
  mask?: Mask;

  /**
   * The maximum value of a valid {@link Input} value.
   */
  max?: ValidationRule<number | string>;

  /**
   * The maximum length of a valid {@link Input} value.
   */
  maxLength?: ValidationRule<number>;

  /**
   * Whether the {@link maxLength} validation rule should limit the number of characters typed by the user.
   *
   * @default false
   */
  maxLengthLimitTyping?: boolean;

  /**
   * The minimum value of a valid {@link Input} value.
   */
  min?: ValidationRule<number | string>;

  /**
   * The minimum length of a valid {@link Input} value.
   */
  minLength?: ValidationRule<number>;

  /**
   * The pattern to match the {@link Input} value against.
   */
  pattern?: ValidationRule<RegExp>;

  /**
   * The {@link Input} reference.
   */
  ref?: Ref<InputRefType>;

  /**
   * The required validation rule of the {@link Input}.
   * If given a non-empty `string`, the {@link Input} will be required and the string will be used as the error message.
   *
   * @default `${label} is required`
   */
  required?: Message | ValidationRule<boolean>;

  /**
   * The custom validation callback function to use for the {@link Input}.
   */
  validate?: ValidateFn<TFieldValues, TFieldName>;

}

/**
 * The {@link Input} component style properties.
 */
export type InputStyleProps = Pick<
  InputProps,
  'containerStyle' | 'disabledInputStyle' | 'errorStyle' | 'inputContainerStyle' | 'inputStyle'
  | 'labelStyle' | 'leftIconContainerStyle' | 'rightIconContainerStyle' | 'style'
>;

/**
 * The {@link Input} functional component type.
 *
 * @template TFieldValues The type of the form data.
 * @template TContext The type of the form context.
 * @template TFieldName The form field name.
 */
export type InputFC = <
  TFieldValues extends FieldValues = any,
  TContext = any,
  TFieldName extends FieldPath<TFieldValues> = Path<TFieldValues>
>(
  props: InputProps<TFieldValues, TContext, TFieldName>,
) => ReactNode;

/**
 * The {@link Input} component reference type.
 */
export type InputRefType = TextInput & PropsWithChildren<RneInputProps>;
