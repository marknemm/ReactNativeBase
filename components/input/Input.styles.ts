import { useGeneralStyles, useThemedStyles } from '@hooks/styles-hooks';
import { StyleSheet } from 'react-native';
import type Input from './Input';
import type { InputStyleProps } from './Input.interfaces';

/**
 * Gets the styles for the {@link Input} component.
 *
 * @param props The component {@link InputStyleProps}.
 * @returns The styles for the {@link Input} component.
 */
export function useStyles({ containerStyle, inputContainerStyle, inputStyle, labelStyle, style }: InputStyleProps) {
  const generalStyles = useGeneralStyles();

  return useThemedStyles(() => ({
    container: {
      ...generalStyles.form.fieldContainer,
      ...StyleSheet.flatten(containerStyle),
    },
    inputContainer: {
      ...generalStyles.form.fieldBorder,
      ...StyleSheet.flatten(inputContainerStyle),
    },
    input: {
      ...generalStyles.form.field,
      ...StyleSheet.flatten(inputStyle),
    },
    label: {
      ...generalStyles.form.label,
      ...StyleSheet.flatten(labelStyle),
    },
    style: {
      ...generalStyles.form.fieldText,
      ...StyleSheet.flatten(style),
    },
  }), [containerStyle, inputContainerStyle, inputStyle, generalStyles, labelStyle, style]);
}
