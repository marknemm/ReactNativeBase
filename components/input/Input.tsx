import { useFormControl, useFormErrorMessage, useValidationRules } from '@hooks/form-hooks';
import { useCallbacks } from '@hooks/state-hooks';
import { Input as RneInput, useTheme } from '@rneui/themed';
import { forwardRef, useEffect, useRef, useState } from 'react';
import { Controller } from 'react-hook-form';
import { useMaskedInputProps } from 'react-native-mask-input';
import type { InputFC, InputProps, InputRefType } from './Input.interfaces';
import { useStyles } from './Input.styles';

/**
 * Component for generic user text input.
 *
 * @param props The component {@link InputProps}.
 * @param ref The component reference.
 * @returns The {@link Input} component.
 */
const Input: InputFC = forwardRef((props, ref) => {
  const { label, name, onBlur, onChangeText } = props;

  // Derive entities related to an input controlled by react-hook-form
  const control = useFormControl(props);
  const rules = useValidationRules(props, label?.toString());
  const errorMessage = useFormErrorMessage(rules, props);

  return control
    ? (
      <Controller
        control={control}
        name={name}
        render={({ field: { onChange: onChangeForm, onBlur: onBlurForm, value } }) => (
          <InputControlled
            {...props}
            errorMessage={errorMessage}
            onBlur={(event) => {
              onBlurForm();
              onBlur?.(event);
            }}
            onChangeText={(text) => {
              onChangeForm(text);
              onChangeText?.(text);
            }}
            ref={ref}
            value={value}
          />
        )}
        rules={rules}
      />
    )
    : <InputControlled {...props} ref={ref} />;
});

/**
 * Controlled {@link Input} component.
 *
 * @param props The component {@link InputProps}.
 * @param ref The component reference.
 * @returns The {@link InputControlled} component.
 */
const InputControlled: InputFC = forwardRef(({
  containerStyle,
  inputContainerStyle,
  inputStyle,
  labelStyle,
  mask,
  maxLength,
  maxLengthLimitTyping,
  onChangeText,
  style,
  value,
  ...inputProps
}, ref) => {
  const styles = useStyles({ containerStyle, inputContainerStyle, inputStyle, labelStyle, style });
  const { theme } = useTheme();
  const [uiValue, setUiValue] = useState('');
  const inputRef = useRef<InputRefType>();
  ref ??= inputRef;

  const maskedInputProps = useMaskedInputProps({ mask, onChangeText, value });
  if (mask) {
    onChangeText = maskedInputProps.onChangeText;
    value = maskedInputProps.value;
  }

  const maxLengthNum = (typeof maxLength === 'number')
    ? maxLength
    : maxLength?.value ?? undefined;

  useEffect(() => {
    // Sync value prop with input (UI) value. Done manually to prevent change on each keystroke and prevent lag.
    if (value !== undefined && value !== uiValue) {
      const textInput = (ref as React.MutableRefObject<InputRefType>).current;
      textInput?.setNativeProps({ text: value });
    }
  }, [ref, value, uiValue]);

  return (
    <RneInput
      keyboardAppearance={theme.mode}
      placeholderTextColor={theme.colors.placeholder}
      {...inputProps}
      {...(mask ? maskedInputProps : undefined)}
      containerStyle={styles.container}
      inputContainerStyle={styles.inputContainer}
      inputStyle={styles.input}
      labelStyle={styles.label}
      maxLength={maxLengthLimitTyping ? maxLengthNum : undefined}
      onChangeText={useCallbacks(onChangeText, setUiValue)}
      ref={ref}
      style={styles.style}
      value={undefined} // Do not update value prop directly to prevent input lag (see useEffect above).
    />
  );
});

export type * from './Input.interfaces';
export default Input;
