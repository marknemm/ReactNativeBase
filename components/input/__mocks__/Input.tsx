import type { InputProps } from '@components/input/Input.interfaces';

const MockInput = jest.fn().mockImplementation((props: InputProps) => {
  const ActualComponent = jest.requireActual('@components/input/Input').default;
  return <ActualComponent {...props} />;
});

export default MockInput;
