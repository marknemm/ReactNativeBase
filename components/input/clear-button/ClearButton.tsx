import type Input from '@components/input/Input';
import { CLEAR_ICON } from '@constants/icons';
import { useColors } from '@hooks/theme-hooks';
import { Button } from '@rneui/themed';
import { forwardRef } from 'react';
import type { ClearButtonProps } from './ClearButton.interfaces';

/**
 * A button that clears a text {@link Input} field.
 */
const ClearButton: React.FC<ClearButtonProps> = forwardRef(({
  inputRef,
  isVisible,
  onPress,
  ...buttonProps
}, ref) => {
  const colors = useColors();

  return isVisible && (
    <Button
      accessibilityLabel="Clear"
      icon={{
        ...CLEAR_ICON,
        color: colors.placeholder,
        size: 18,
      }}
      onPress={(event) => {
        onPress?.(event);
        if (!event?.isDefaultPrevented() && !event?.defaultPrevented) {
          inputRef.current?.clear();
          inputRef.current?.props.onChangeText?.('');
        }
      }}
      size="sm"
      type="clear"
      {...buttonProps}
      ref={ref}
    />
  );
});

export type * from './ClearButton.interfaces';
export default ClearButton;
