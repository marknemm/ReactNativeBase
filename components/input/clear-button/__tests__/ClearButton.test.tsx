import { type InputRefType } from '@components/input/Input';
import ClearButton from '@components/input/clear-button/ClearButton';
import AppProvider from '@test/contexts/app/AppProvider';
import { render, screen, userEvent } from '@testing-library/react-native';
import { type RefObject } from 'react';
import { GestureResponderEvent } from 'react-native';

describe('<ClearButton />', () => {
  let mockInputRef: RefObject<InputRefType>;

  beforeEach(() => {
    mockInputRef = { current: { clear: jest.fn(), props: { onChangeText: jest.fn() } } as any };
  });

  describe('button press', () => {
    it('calls onPress and clears input when the button is pressed', async () => {
      const onPress = jest.fn();
      render(
        <ClearButton
          inputRef={mockInputRef}
          isVisible
          onPress={onPress}
        />,
        { wrapper: AppProvider }
      );

      const clearButton = screen.getByRole('button', { name: 'Clear' });
      await userEvent.press(clearButton);

      expect(onPress).toHaveBeenCalled();
      expect(mockInputRef.current.clear).toHaveBeenCalled();
      expect(mockInputRef.current.props.onChangeText).toHaveBeenCalledWith('');
    });

    it('does not call onPress or clear input when disabled', async () => {
      const onPress = jest.fn();
      render(
        <ClearButton
          disabled
          inputRef={mockInputRef}
          isVisible
          onPress={onPress}
        />,
        { wrapper: AppProvider }
      );

      const clearButton = screen.getByRole('button', { name: 'Clear' });
      await userEvent.press(clearButton);

      expect(onPress).not.toHaveBeenCalled();
      expect(mockInputRef.current.clear).not.toHaveBeenCalled();
      expect(mockInputRef.current.props.onChangeText).not.toHaveBeenCalled();
    });

    it('does clear input when onPress prevents default event', async () => {
      const onPress = jest.fn().mockImplementation((event: GestureResponderEvent) => {
        event.preventDefault();
        event.defaultPrevented = true;
      });
      render(
        <ClearButton
          inputRef={mockInputRef}
          isVisible
          onPress={onPress}
        />,
        { wrapper: AppProvider }
      );

      const clearButton = screen.getByRole('button', { name: 'Clear' });
      await userEvent.press(clearButton);

      expect(onPress).toHaveBeenCalled();
      expect(mockInputRef.current.clear).not.toHaveBeenCalled();
      expect(mockInputRef.current.props.onChangeText).not.toHaveBeenCalledWith('');
    });
  });

  describe('snapshots', () => {
    it('renders correctly when not visible', () => {
      render(
        <ClearButton inputRef={mockInputRef} />,
        { wrapper: AppProvider }
      );

      expect(screen.toJSON()).toMatchSnapshot();
    });

    it('renders correctly when visible', () => {
      render(
        <ClearButton
          inputRef={mockInputRef}
          isVisible
        />,
        { wrapper: AppProvider }
      );

      expect(screen.toJSON()).toMatchSnapshot();
    });

    it('renders correctly when visible and disabled', () => {
      render(
        <ClearButton
          disabled
          inputRef={mockInputRef}
          isVisible
        />,
        { wrapper: AppProvider }
      );

      expect(screen.toJSON()).toMatchSnapshot();
    });
  });
});
