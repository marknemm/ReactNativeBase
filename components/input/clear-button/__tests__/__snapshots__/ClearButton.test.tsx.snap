// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ClearButton /> snapshots renders correctly when not visible 1`] = `null`;

exports[`<ClearButton /> snapshots renders correctly when visible 1`] = `
<View
  style={
    [
      {
        "overflow": "hidden",
      },
      {
        "borderRadius": 3.142857142857143,
      },
      undefined,
      false,
    ]
  }
  testID="RNE_BUTTON_WRAPPER"
>
  <View
    accessibilityLabel="Clear"
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": false,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "borderColor": "#007aff",
          "borderRadius": 3.142857142857143,
          "borderWidth": 0,
          "flexDirection": "row",
          "justifyContent": "center",
          "padding": 6.285714285714286,
          "paddingHorizontal": 8.285714285714285,
        }
      }
    >
      <View
        style={
          {
            "marginHorizontal": 5,
            "overflow": "hidden",
          }
        }
        testID="RNE__ICON__CONTAINER"
      >
        <View
          accessibilityRole="button"
          android_ripple={
            {
              "borderless": false,
              "color": "rgba(0, 0, 0, 0.3)",
              "radius": -5,
            }
          }
          disabled={false}
          testID="RNE__ICON__CONTAINER_ACTION"
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "justifyContent": "center",
              }
            }
            testID="RNE__ICON"
          >
            <Text
              allowFontScaling={false}
              selectable={false}
              style={
                [
                  {
                    "color": "#86939e",
                    "fontSize": 18,
                  },
                  {
                    "backgroundColor": "transparent",
                  },
                  {
                    "fontFamily": "material",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  {},
                ]
              }
              testID="RNE__ICON__Component"
            >
              
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`<ClearButton /> snapshots renders correctly when visible and disabled 1`] = `
<View
  style={
    [
      {
        "overflow": "hidden",
      },
      {
        "borderRadius": 3.142857142857143,
      },
      undefined,
      false,
    ]
  }
  testID="RNE_BUTTON_WRAPPER"
>
  <View
    accessibilityLabel="Clear"
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": false,
        "checked": undefined,
        "disabled": true,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "borderColor": "#007aff",
          "borderRadius": 3.142857142857143,
          "borderWidth": 0,
          "flexDirection": "row",
          "justifyContent": "center",
          "padding": 6.285714285714286,
          "paddingHorizontal": 8.285714285714285,
        }
      }
    >
      <View
        style={
          {
            "marginHorizontal": 5,
            "overflow": "hidden",
          }
        }
        testID="RNE__ICON__CONTAINER"
      >
        <View
          accessibilityRole="button"
          android_ripple={
            {
              "borderless": false,
              "color": "rgba(0, 0, 0, 0.3)",
              "radius": -5,
            }
          }
          disabled={false}
          testID="RNE__ICON__CONTAINER_ACTION"
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "justifyContent": "center",
              }
            }
            testID="RNE__ICON"
          >
            <Text
              allowFontScaling={false}
              selectable={false}
              style={
                [
                  {
                    "color": "#86939e",
                    "fontSize": 18,
                  },
                  {
                    "backgroundColor": "transparent",
                  },
                  {
                    "fontFamily": "material",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  {},
                ]
              }
              testID="RNE__ICON__Component"
            >
              
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
